### PR TITLE
Add build option to enable cuda profiling

### PR DESCRIPTION
**Pull request by RandySheriffH**
Make CUDA profiling functions a build option, by default disabled.

### CODE REVIEW

File: tools/ci_build/build.py [hunk 1]
```diff
@@ -567,6 +567,9 @@ def parse_arguments():
         "--test_external_transformer_example", action='store_true',
         help="run the example external transformer test, mainly used in CI pipeline.")
 
+    parser.add_argument(
+        "--enable_cuda_profiling", action='store_true', help="enable cuda kernel profiling.")
```
Comment (pranavsharma) [inline tools/ci_build/build.py:571]: Can add a note here saying this requires ensuring cupti is in the PATH.

**Review by pranavsharma — COMMENTED** [review]
Some changes from this https://github.com/microsoft/onnxruntime/pull/9836/files need to be reverted as well.

File: onnxruntime/core/providers/cuda/cuda_profiler.cc [hunk 2]
```diff
@@ -1,6 +1,6 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
-#if !(defined(USE_ROCM) || defined(ENABLE_TRAINING))
+#if defined(ENABLE_CUDA_PROFILING) && !(defined(USE_ROCM) || defined(ENABLE_TRAINING))
```
Comment (jywu-msft) [inline onnxruntime/core/providers/cuda/cuda_profiler.cc:3]: where is ENABLE_CUDA_PROFILING defined?

File: onnxruntime/test/framework/inference_session_test.cc [hunk 3]
```diff
@@ -661,7 +661,7 @@ TEST(InferenceSessionTests, CheckRunProfilerWithSessionOptions) {
     }
   }
 
-#if defined(USE_CUDA) && !defined(ENABLE_TRAINING) && defined(CUDA_VERSION) && CUDA_VERSION >= 11000
+#if defined(USE_CUDA) && ENABLE_CUDA_PROFILING && !defined(ENABLE_TRAINING) && defined(CUDA_VERSION) && CUDA_VERSION >= 11000
```
Comment (snnn) [inline onnxruntime/test/framework/inference_session_test.cc:664]: Let's keep it consistent.

File: onnxruntime/core/providers/cuda/cuda_profiler.cc (hunk 2 above)
Comment (jywu-msft) [reply onnxruntime/core/providers/cuda/cuda_profiler.cc:3]: it needs to be defined when onnxruntime_ENABLE_CUDA_PROFILING is defined right?

Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/cuda_profiler.cc:3]: It is defined in CMakeList.txt added in latest commit.

File: onnxruntime/test/framework/inference_session_test.cc (hunk 3 above)
Comment (RandySheriffH) [reply onnxruntime/test/framework/inference_session_test.cc:664]: Sure, done.

File: cmake/CMakeLists.txt [hunk 4]
```diff
@@ -175,6 +175,8 @@ option(onnxruntime_PREBUILT_PYTORCH_PATH "Path to pytorch installation dir")
 # external transformer src path
 option(onnxruntime_EXTERNAL_TRANSFORMER_SRC_PATH "Path to external transformer src dir")
 
+option(onnxruntime_ENABLE_CUDA_PROFILING "Eanble CUDA kernel profiling" OFF)
```
Comment (pranavsharma) [inline cmake/CMakeLists.txt:178]:
```suggestion
option(onnxruntime_ENABLE_CUDA_PROFILING "Enable CUDA kernel profiling" OFF)
```

File: cmake/CMakeLists.txt [hunk 5]
```diff
@@ -1700,6 +1702,10 @@ if (onnxruntime_ENABLE_TRAINING_OPS)
   add_compile_definitions(ENABLE_TRAINING_OPS)
 endif()
 
+if (onnxruntime_ENABLE_CUDA_PROFILING)
```
Comment (jywu-msft) [inline cmake/CMakeLists.txt:1705]: shouldn't you also check onnxruntime_ENABLE_CUDA as well?
or stick this check underneath
https://github.com/microsoft/onnxruntime/blob/ce84554ed0084150221499a7d6b5b439ce7cffba/cmake/CMakeLists.txt#L1571

File: cmake/onnxruntime_providers.cmake [hunk 6]
```diff
@@ -354,12 +354,18 @@ if (onnxruntime_USE_CUDA)
 
   add_dependencies(onnxruntime_providers_cuda onnxruntime_providers_shared ${onnxruntime_EXTERNAL_DEPENDENCIES} ${onnxruntime_tvm_dependencies})
   target_link_directories(onnxruntime_providers_cuda PRIVATE ${onnxruntime_CUDA_HOME}/extras/CUPTI/lib64)
-  target_link_libraries(onnxruntime_providers_cuda PRIVATE cublas cudnn curand cufft cupti ${ONNXRUNTIME_PROVIDERS_SHARED})
-  target_include_directories(onnxruntime_providers_cuda PRIVATE ${ONNXRUNTIME_ROOT} ${CMAKE_CURRENT_BINARY_DIR} ${onnxruntime_CUDNN_HOME}/include ${eigen_INCLUDE_DIRS} ${TVM_INCLUDES} PUBLIC ${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES} ${onnxruntime_CUDA_HOME}/extras/CUPTI/include)
+
```
Comment (jywu-msft) [inline cmake/onnxruntime_providers.cmake:357]: doesn't line 356 need to be moved ? 
target_link_directories(onnxruntime_providers_cuda PRIVATE ${onnxruntime_CUDA_HOME}/extras/CUPTI/lib64)

**Review by pranavsharma — DISMISSED** [review]

File: tools/ci_build/build.py [hunk 7]
```diff
@@ -567,6 +567,9 @@ def parse_arguments():
         "--test_external_transformer_example", action='store_true',
         help="run the example external transformer test, mainly used in CI pipeline.")
 
+    parser.add_argument(
+        "--enable_cuda_profiling", action='store_true', help="enable cuda kernel profiling, cupti library must be added to PATH beforehand.")
```
Comment (jywu-msft) [inline tools/ci_build/build.py:571]: this line is too long. fails FLAKE8 check.

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**